keps/sig-api-machinery/20180731-crd-pruning.md

```diff
@@ -0,0 +1,610 @@
+   ```
+
+   Pruning takes place, but `apiVersion`, `kind`, `metadata` and known fields under `metadata` are preserved, i.e.
+   
+   ```json
+   {
+     "apiVersion": "example/v1",
+     "kind": "Foo",
+     "metadata": {
+       "name": "example",
+       "garbage": 43
+     },
+     "foo": 42
+   }
+   ```
+   
+   is pruned to
+   
+   ```json
+   {
+     "apiVersion": "example/v1",
+     "kind": "Foo",
+     "metadata": {
+       "name": "example"
+     }
+   }
+   ```
+
+### Opt-in and Opt-out of Pruning on CRD Level
+
+We will add a `PruneUnknownFields` flag to `CustomResourceDefinitionSpec` of `apiextensions.k8s.io/v1beta1`:
+
+```go
+type CustomResourceDefinitionSpec struct {
+  ...
+	
+  // pruneUnknownFields enables pruning of object fields which are not
+  // specified in the OpenAPI schema. apiVersion, kind, metadata and known
+  // fields inside metadata are excluded from pruning.
+  // Defaults to false. 
+  // Setting this field to true is considered an alpha API.
+  // Note: this will default to true in version v1.
+  PruneUnknownFields bool
+}
+```
+
+I.e. for `apiextensions.k8s.io/v1beta1` this will default to `false` for backwards compatibility.
+
+For `apiextensions.k8s.io/v1` we will change the default to `true` and forbid `false` during creation and updates if it has been `true` before. In `v1` the only way to opt-out from pruning is via setting `x-kubernetes-preserve-unknown-fields: true` in the schema. An empty or unspecified schema will prune everything.
+
+### References
+
+* Old pruning implementation PR https://github.com/kubernetes/kubernetes/pull/64558, to be adapted 
+* [OpenAPI v3 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md)
+* [JSON Schema](http://json-schema.org/)
+
+### Test Plan
+
+**blockers for alpha:**
+
+We default `PruneUnknownFields` to false and hence switch off the whole code path doing pruning. This reduces risk for everybody not using this alpha feature.
+
+* we add unit tests for the general pruning algorithm
+* we add apiextensions-apiserver integration tests to
+  * verify that the pruning feature is actually off if `PruneUnknownFields` is false.
+  * verify that `PruneUnknownFields`  is defaulted to false.
+  * verify that pruning happens if `PruneUnknownFields` is true, for all versions in the CRD according to the schema of the respective version.
+  * verify that `metadata`, `apiVersion`, `kind` are preserved if `PruneUnknownFields` is true and there is no schema given in the CRD.
+  
+**blockers for beta:**
+
+* we implement and verify that `x-kubernetes-embedded-resource` and `x-kubernetes-preserve-unknown-fields` work as expected.
+* we add apiextensions-apiserver integration tests to
+  * verify that pruning happens on incoming request payloads, on read from storage and after calling mutating admission webhooks.
+
+**blockers for GA:**
+
+* we verified that performance of pruning is adequate and not considerably reducing throughput.
```

Comment:
this needs more detail, but can be fleshed out as we plan the release where this gets promoted to GA

Reply:
Also I would wait for the GA KEP to settle on requirements. I expect this to be considerably cheaper than decoding. In the common case of very few pruned fields, this should be nearly without allocations, just a double-recursion.